### PR TITLE
rg-68: Email unsubscribe

### PR DESCRIPTION
**Pull request by honkster-nazarii**
Finished, depends on #107 
---
Subscribing
---
If user is subscribed, he will have this property:
зображення
This will affect subscribe button state in user profile settings

Unsubscribing
---
User can unsubscribe by hitting
/email-subscription/unsubscribe/:email_subscription_id
зображення
This link should be added to emails (when they are done)

Demo:

https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-resumegemm/assets/133093930/2e36e95f-6b1b-4b7c-9bc4-71b5f9c7487a

### CODE REVIEW

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

**Review by MaxPoplavskiy — APPROVED** [review]

File: backend/src/bundles/email-subscriptions/email-subscription.controller.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        }>,
+    ): Promise<ApiHandlerResponse<void>> {
+        try {
+            await this.emailSubscriptionService.unsubscribe(options.params.id);
+            return {
+                status: HttpCode.OK,
+                payload: {
+                    message: 'Successfully unsubscribed',
+                },
+            };
```
Comment (Icerock10) [inline backend/src/bundles/email-subscriptions/email-subscription.controller.ts:147]: Maybe an Exception message for that?

**Comment by Icerock10** [comment]
conflicts

**Review by Icerock10 — APPROVED** [review]

File: backend/src/bundles/email-subscriptions/email-subscription.repository.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+class EmailSubscriptionRepository implements IEmailSubscriptionRepository {
+    private emailSubscriptionModel: typeof EmailSubscriptionModel;
+
+    public constructor(emailSubscriptionModel: typeof EmailSubscriptionModel) {
+        this.emailSubscriptionModel = emailSubscriptionModel;
+    }
+
+    public async create(transaction: Transaction): Promise<EmailSubscription> {
+        return await this.emailSubscriptionModel
```
Comment (Icerock10) [inline backend/src/bundles/email-subscriptions/email-subscription.repository.ts:18]: return await, you have multiple such lines in code, you may adjust

File: backend/src/bundles/email-subscriptions/email-subscription.service.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            });
+        }
+
+        const transaction = await this.userRepository.model.startTransaction();
+
+        try {
+            const subscription = await this.emailSubscriptionRepository.create(
+                transaction,
+            );
+            void (await this.userRepository.updateEmailSubscriptionId(
```
Comment (Icerock10) [inline backend/src/bundles/email-subscriptions/email-subscription.service.ts:32]: does void influence the behavior?

File: frontend/src/bundles/users/components/subscription/subscriptions.tsx [hunk 4]
```diff
@@ -1,9 +1,76 @@
...
+    unsubscribe,
+} from '~/bundles/email-subscription/store/actions';
+import { ToastType } from '~/bundles/toast/enums/show-toast-types.enum';
+import { showToast } from '~/bundles/toast/helpers/show-toast';
+
 import styles from './style.module.scss';
 import { SubscriptionItem } from './subscription-item';
+import { SubscriptionToggleItem } from './subscription-toggle-item';
 
 const Subscriptions: React.FC = () => {
```
Comment (Icerock10) [inline frontend/src/bundles/users/components/subscription/subscriptions.tsx:21]: Don't you want to separate the business logic let's say in custom hook and keep the component as `presentational`?
As you wish

**Review by Icerock10 — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/users/components/subscription/subscriptions.tsx (hunk 4 above)
Comment (honkster-nazarii) [reply frontend/src/bundles/users/components/subscription/subscriptions.tsx:14]: is it ok as it is now?

Comment (Icerock10) [reply frontend/src/bundles/users/components/subscription/subscriptions.tsx:14]: You can move everything there, and return what your component requires, for example like so, and prettify it inside custom hook

```
const {
        handleSubscribe,
        handleUnsubscribe,
        isEmailSubscriptionLoading,
        toggleItemProperties,
        isModalOpen,
        handleModalClose,
        handleUnsubscribeAndCloseModal,
    } = useEmailSubscriptions();
```

File: frontend/src/bundles/common/hooks/use-email-subscriptions/use-email-subscriptions.hook.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import { useCallback, useState } from 'react';
```
Comment (Icerock10) [inline frontend/src/bundles/common/hooks/use-email-subscriptions/use-email-subscriptions.hook.ts:1]:
```suggestion
import { useCallback, useState, useAppDispatch, useAppSelector } from '../hooks';
```

**Review by Icerock10 — APPROVED** [review]

File: frontend/src/bundles/common/hooks/use-email-subscriptions/use-email-subscriptions.hook.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+                        ToastType.SUCCESS,
+                    );
+                    handleModalClose();
+                });
+        }
+    }, [dispatch, handleModalClose, user]);
+
+    const toggleItemProperties = user?.emailSubscription
+        ? {
+              title: 'Email notifications',
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/common/hooks/use-email-subscriptions/use-email-subscriptions.hook.ts:76]: Move duplicate values into variables

File: frontend/src/bundles/users/components/subscription/subscription-toggle-item.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        <div className={styles.subscription__item}>
+            <div className={styles.subscription__item__content}>
+                <p>
+                    <strong>{title}</strong>
+                </p>
+                <p>{info}</p>
+            </div>
+            <div>
+                <RegularButton onClick={onClick}>
+                    {isLoading ? 'Loading...' : buttonText}
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/users/components/subscription/subscription-toggle-item.tsx:30]: Check button component, it should have a loading state

File: frontend/src/bundles/email-subscription/pages/email-subscription-page.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    useEffect(() => {
+        if (parameters.id) {
+            void dispatch(unsubscribe({ id: parameters.id }));
+        }
+    }, [dispatch, parameters.id]);
+
+    return (
+        <div className={styles.email_subscription}>
+            {isEmailSubscriptionLoading === DataStatus.PENDING ? (
+                <p>Wait...</p>
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/email-subscription/pages/email-subscription-page.tsx:29]: Maybe it worth adding spinner here?

File: frontend/src/bundles/users/components/subscription/subscriptions.tsx [hunk 9]
```diff
@@ -20,6 +34,32 @@ const Subscriptions: React.FC = () => {
...
+                    <div className={styles.modal__buttons}>
+                        <RegularButton onClick={handleModalClose}>
+                            Cancel
+                        </RegularButton>
+                        <RegularButton
+                            onClick={handleUnsubscribe}
+                            variant={ButtonVariant.PRIMARY}
+                        >
+                            {isEmailSubscriptionLoading
+                                ? 'Loading...'
```
Comment (Vitaliy-Kalenichenko) [inline frontend/src/bundles/users/components/subscription/subscriptions.tsx:57]: Same here

**Review by Vitaliy-Kalenichenko — CHANGES_REQUESTED** [review]

**Comment by honkster-nazarii** [comment]
this pr is cursed with conflicts

**Pull request closed without merging**